jwql/shared_tasks/shared_tasks.py

```diff
@@ -0,0 +1,806 @@
+
+
+def run_parallel_pipeline(input_files, in_ext, ext_or_exts, instrument, jump_pipe=False):
+    """Convenience function for using the ``run_calwebb_detector1`` function on a list of
+    data files, breaking them into parallel celery calls, collecting the results together,
+    and returning the results as another list. In particular, this function will do the
+    following:
+    
+    - Lock the file ID so that no other calibration happens at the same time
+    - Copy the input (raw) file to the (central storage) transfer location
+    - Call the ``run_calwebb_detector1`` task
+    - For the extension (or extensions) (where by "extension" we mean 'uncal' or 'refpix'
+      or 'jump' rather than something like '.fits') requested, copy the files from the 
+      outgoing transfer location to the same directory as the input file
+    - Delete the input file from the transfer location
+    - Delete the output files from the transfer location
+    
+    It will then return what it was given – either a single file+path or a list of 
+    files+paths, depending on what ``out_exts`` was provided as.
+
+    Parameters
+    ----------
+    input_file : str
+        Name of fits file to run on the pipeline
+    
+    in_ext : str
+        Input file extension
+
+    ext_or_exts : str or list-of-str or dict
+        The requested output calibrated files. This must be either:
+        
+        - A string indicating a single extension to be retrieved for all files.
+        - A list of strings indicating multiple extensions to be retrieved for all files.
+        - A dict with a key for each input file, containing either a single extension 
+          string or a multiple-extension list of strings to be retrieved for that file
+          (note that a default dict can be used here)
+    
+    instrument : str
+        Name of the instrument being calibrated
+    
+    jump_pipe : bool
+        Whether the detector1 jump pipeline is being used (e.g. the bad pixel monitor)
+
+    Returns
+    -------
+    file_or_files : str or list-of-str
+        Name (or names) of the result file(s), including path(s)
+    """
+    logging.info("Pipeline call requestion calibrated extensions {}".format(ext_or_exts))
+    for input_file in input_files:
+        logging.info("\tCalibrating {}".format(input_file))
+    
+    input_file_paths = {}
+    results = {}
+    locks = {}
+    outputs = {}
+    output_dirs = {}
+
+    logging.info("Dispatching celery tasks")
+    try:
+        for input_file in input_files:
+            retrieve_dir = os.path.dirname(input_file)
+            logging.info("\tPipeline call for {} requesting {} sent to {}".format(input_file, ext_or_exts, retrieve_dir))
+            short_name, cal_lock, uncal_file = prep_file(input_file, in_ext)
+            uncal_name = os.path.basename(uncal_file)
+            output_dirs[short_name] = retrieve_dir
+            input_file_paths[short_name] = input_file
+            locks[short_name] = cal_lock
```

Comment:
Effectively yes. In particular, the function `prep_file()`, which returns the "cal_lock" instance, sets up the lock (and does so in a blocking way, meaning that if the lock isn't available the function will wait until it *is* available, and has been acquired, before returning.

With using redis for locking, there are essentially two ways you can run a lock call:

- `blocking=True` will keep the function from completing until the lock is available and has been acquired. In this case, when you call the lock's `have_lock()` method, it should *always* return True.
- `blocking=False` will return right away. In this case, you need to check whether you were actually *given* the lock by calling its `have_lock()` method. If it returns True, you are the lock holder. If it returns False, then you are not the lock holder. Note that the `@only_one()` decorator uses a non-blocking call, and so if you try to run a monitor when the lock is held by another copy of the same monitor, then the monitor's `process()` function simply won't run, and what I have them doing now is to print out a message saying that the lock is in use, and then quit.

Now, because (currently) the pipeline file locks *never* time out, you have to release them when you're done (which this function always does after it's finished all of the calibrations). It's possible that it could, instead, release the lock as soon as it's finished the calibration, which might help with reprocessing if you end up having contention between various different monitors that want to calibrate the same set of files. Also, if you try to release a lock that you don't have, it raises an exception. Currently that's unlikely to happen, but it's also not harmful if it does, so quite possibly we could safely catch-log-and-ignore that exception, but I don't think it's an urgent question.

Reply:
Cool, thanks for the explanation. I will follow up on slack if I have more questions regarding this topic.